lib/spack/spack/solver/concretize.lp

```diff
@@ -115,29 +115,51 @@ multiple_nodes_attribute("depends_on").
 multiple_nodes_attribute("virtual_on_edge").
 multiple_nodes_attribute("provider_set").
 
-% Map constraint on the literal ID to facts on the node
-attr(Name, node(min_dupe_id, A1))             :- literal(LiteralID, Name, A1), solve_literal(LiteralID).
-attr(Name, node(min_dupe_id, A1), A2)         :- literal(LiteralID, Name, A1, A2), solve_literal(LiteralID), not multiple_nodes_attribute(Name).
-attr(Name, node(min_dupe_id, A1), A2, A3)     :- literal(LiteralID, Name, A1, A2, A3), solve_literal(LiteralID), not multiple_nodes_attribute(Name).
-attr(Name, node(min_dupe_id, A1), A2, A3, A4) :- literal(LiteralID, Name, A1, A2, A3, A4), solve_literal(LiteralID).
-
-% Special cases where nodes occur in arguments other than A1
-attr("node_flag_source", node(min_dupe_id, A1), A2, node(min_dupe_id, A3)) :- literal(LiteralID, "node_flag_source", A1, A2, A3), solve_literal(LiteralID).
-attr("depends_on",       node(min_dupe_id, A1), node(min_dupe_id, A2), A3) :- literal(LiteralID, "depends_on",       A1, A2, A3), solve_literal(LiteralID).
-
-attr("virtual_node", node(min_dupe_id, Virtual))                              :- literal(LiteralID, "provider_set", _,        Virtual), solve_literal(LiteralID).
-attr("provider_set", node(min_dupe_id, Provider), node(min_dupe_id, Virtual)) :- literal(LiteralID, "provider_set", Provider, Virtual), solve_literal(LiteralID).
-provider(node(min_dupe_id, Provider), node(min_dupe_id, Virtual))             :- literal(LiteralID, "provider_set", Provider, Virtual), solve_literal(LiteralID).
+trigger_condition_holds(TriggerID, node(min_dupe_id, Package)) :-
+  solve_literal(TriggerID),
+  pkg_fact(Package, condition_trigger(_, TriggerID)),
+  literal(TriggerID).
+
+trigger_node(TriggerID, Node, Node) :-
+  trigger_condition_holds(TriggerID, Node),
+  literal(TriggerID).
+
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, _, Mentioned, _, _, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, "depends_on", _, _, Mentioned, _).
+mentioned_in_literal(Root, Mentioned) :-
+  literal(TriggerID),
+  trigger_and_effect(Root, TriggerID, EffectID),
+  imposed_constraint(EffectID, "provider_set", _, Mentioned).
+
+condition_set(node(min_dupe_id, Root), node(min_dupe_id, Mentioned), link_run) :-
+  mentioned_in_literal(Root, Mentioned).
```

Comment:
I think this is wrong. For instance with a spec like:
```
hdf5 ^cmake@3.26
```
`cmake` is not a link-run dependency in the condition set of `hdf5`. Also, I should have deleted `link_run` since we are using only `direct_link_run` and `direct_build`.

Reply:
See https://github.com/spack/spack/pull/40173/commits/67dd6cebd5557a17647928c0e0122c7dce0049db (I have to think hard if there are edge cases where computing `/2` facts directly for roots may be wrong).